src/main/scala/is/hail/io/vcf/ExportVCF.scala

```diff
@@ -190,9 +188,27 @@ object ExportVCF {
     }(sb += ':')
   }
 
+  def getAttributes(path: List[String], md: Option[Map[String, Any]]): Option[Any] = {
+    if (path.isEmpty)
+      None
+    else {
+      if (path.length == 1) {
+        md.flatMap(_.get(path.head))
+      } else {
+        md.flatMap { m => m.get(path.head) match {
+          case Some(x: Map[String, Any]) => getAttributes(path.tail, Some(x))
+          case _ => None
+        }
```

Comment:
If you allow me to change the empty path to mean give me all the attributes (instead of `None`), then I think this method gets a lot simpler. It's also a little nicer when using Scala's pattern matching for `List`.

```scala
def getAttributes(path: List[String], attributes: Option[Map[String, Any]]): Option[Any] = path match {
  case Nil =>
    attributes
  case head :: rest =>
    getAttributes(rest, attributes.map(_.get(head).asInstanceOf[Map[String, Any]])
}
```

---

Aside: the error you're getting is because Scala has no way to know if `x` has String keys or not, that information is lost at run-time. To remove the error, you'd have to write:

```scala
case Some(x) => getAttributes(path.tail, Some(x.asInstanceOf[Map[String, Any]]))
```

---

The reason you have to cast is because the structure of your data is not encoded in the Scala type system. AFAICT, you have a triply nested set of keys. Why not use a type alias like this:
```scala
type VcfAttributes = Map[String, Map[String, Map[String, String]]]
```

Then we can rewrite getAttributes as follows:

```scala
def getAttributes(path: List[String], attributes: Option[VcfAttributes]): Option[Any] = path match {
  case Nil =>
    attributes
  case k1 :: Nil =>
    attributes
      .flatMap(_.get(k1))
  case k1 :: k2 :: Nil =>
    attributes
      .flatMap(_.get(k1))
      .flatMap(_.get(k2))
  case k1 :: k2 :: k3 :: Nil => 
    attributes
      .flatMap(_.get(k1))
      .flatMap(_.get(k2))
      .flatMap(_.get(k3))
  case _ => throw new IllegalArgumentException(s"Too many keys! ${path}")
}
```

Reply:
Or really:
```scala
def getAttributes(k1: String, attributes: Option[VcfAttributes]): Option[Map[String, Map[String, String]]] =
  attributes.flatMap(_.get(k1))
def getAttributes(k1: String, k2: String, attributes: Option[VcfAttributes]): Option[Map[String, String]] =
  getAttributes(k1).flatMap(_.get(k2))
def getAttributes(k1: String, k2: String, k3: String, attributes: Option[VcfAttributes]): Option[String] =
  getAttributes(k1, k2).flatMap(_.get(k3))
```

Which saves the string allocation and the matching! And you get the perfectly precise types!